pcdsdevices/lodcm.py

```diff
@@ -60,6 +69,452 @@ def __init__(self, prefix, *args, **kwargs):
+
+    def __init__(self, prefix, *args, **kwargs):
+        self._m_prefix = ''
+        if 'XPP' in prefix:
+            self._m_prefix = 'XPP'
+        elif 'XCS' in prefix:
+            self._m_prefix = 'HFX'
+        super().__init__(prefix, *args, **kwargs)
+
+    def is_diamond(self):
+        """Check if tower 2 is with Diamond (C) material."""
+        return (self.h2n_state.position == 'C' and
+                self.y2_state.position == 'C' and
+                self.chi2_state.position == 'C')
+
+    def is_silicon(self):
+        """Check if tower 2 is with Silicon (Si) material."""
+        return (self.h2n_state.position == 'Si' and
+                self.y2_state.position == 'Si' and
+                self.chi2_state.position == 'Si')
+
+    def get_reflection(self, check=False):
+        """Get crystal's reflection."""
+        reflection = None
+        if self.is_diamond():
+            reflection = self.diamond_reflection.get()
+        elif self.is_silicon():
+            reflection = self.silicon_reflection.get()
+
+        if check and reflection is None:
+            raise ValueError('Unable to determine the crystal reflection')
+        return reflection
+
+    def get_material(self, check=False):
+        """Get the current material."""
+        if self.is_diamond():
+            return 'C'
+        elif self.is_silicon():
+            return 'Si'
+        if check:
+            raise ValueError("Unable to determine crystal material")
+
+
+class DiagnosticsTower(BaseInterface, Device):
+    """
+    Diagnostic Tower Motors.
+
+    Parameters
+    ----------
+    prefix : str
+        Epics base PV prefix.
+    """
+    # Located midway between T1 and T2 in the center of rotation of the device.
+    # horizontal slits
+    dh = Cpt(IMS, ':MON:MMS:16', kind='normal', doc='LOM Dia H')
+    # vertical slits
+    dv = Cpt(IMS, ':MON:MMS:17', kind='normal', doc='LOM Dia V')
+    dr = Cpt(IMS, ':MON:MMS:19', kind='normal', doc='LOM Dia Theta')
+    # filters wheel
+    df = FCpt(IMS, '{self._prefix}:MON:MMS:{self._df_suffix}', kind='normal',
+              doc='LOM Dia Filter Wheel')
+    # pips diode
+    dd = Cpt(IMS, ':MON:MMS:18', kind='normal', doc='LOM Dia PIPS')
+    # yag screen
+    yag_zoom = Cpt(IMS, ':MON:CLZ:01', kind='normal', doc='LOM Zoom')
+
+    tab_component_names = True
+
+    def __init__(self, prefix, *args, **kwargs):
+        self._prefix = prefix
+        self._df_suffix = '27'
+        if 'XCS' in self._prefix:
+            self._df_suffix = '22'
+        super().__init__(prefix, *args, **kwargs)
+
+
+class OffsetIMS(PseudoPositioner):
+    """
+    IMS motor with an offset.
+
+    The driving idea for OffsetIMS is to let us separate the
+    various unrelated PseudoPositioner-like things from the main class.
+    Each different calculation can be independent, making it simpler to
+    implement and maintain.
+    """
+    motor = FCpt(IMS, '{self._prefix}', kind='normal')
```

Comment:
@ZLLentz  and @klauer  - I still don't know if I got this little guy right, can you pretty please help when you get a chance. 

I think what we eventually want to have is simulate the old code:?

```
    def _create_crystal_vm(self, name, motor, offset_pv, use_ims_preset=False):
        offset_pv = "%s:%s" % (self.pvBase, offset_pv)

        def move(pos):
            motor.mv(pos + Pv.get(offset_pv))

        def wm():
            return motor.wm() - Pv.get(offset_pv)

        def set(value):
            if use_ims_preset:
                Pv.put("%s_SET" % offset_pv, motor.wm() - value)
            Pv.put(offset_pv, motor.wm() - value)

        return virtualmotor.VirtualMotor(
            name, move, wm, set=set, motorsobj=self._motors
        )

```
I have a 'REAL' offset motors, that have valid PVs, (`th1_c`, `th1_si`... and so on)

I either don't know how to test this guy properly or I am missing something or have some stuff wrong. 

Here is how I am testing it:

```
In [3]: from pcdsdevices.lodcm import SIMOffsetIMS

In [4]: from ophyd.sim import make_fake_device

In [5]: fake_ims = make_fake_device(SIMOffsetIMS)

In [6]: ims = fake_ims('FAKE:IMS', name='fake_ims')

In [7]: ims
Out[7]: 

fake_ims
--------
position: 0.0

motor
-----
  position: 0

offset
------
  position: 0
  notepad_setpoint: 0
  notepad_readback: 0

In [8]: ims.move(2)
Out[8]: MoveStatus(done=True, pos=fake_ims, elapsed=0.0, success=True, settle_time=0.0)

In [9]: ims
Out[9]: 

fake_ims
--------
position: 0.0

motor
-----
  position: 2

offset
------
  position: 0
  notepad_setpoint: 2
  notepad_readback: 0
```
Now for a real position of 3, I would have to move the offset at 1?
```
In [10]: ims.inverse(3)
Out[10]: FakeSIMOffsetIMSPseudoPos(offset=1)

In [11]: ims.offset.move(1)
Out[11]: MoveStatus(done=True, pos=fake_ims, elapsed=0.0, success=True, settle_time=0.0)

In [12]: ims
Out[12]: 

fake_ims
--------
position: 0.0

motor
-----
  position: 3

offset
------
  position: 0
  notepad_setpoint: 1
  notepad_readback: 0
```
Inverse seems to work, maybe.... although I don't know why the position does not get updated for offset...... 

The `forward` guys is something else though, I am doing something wrong here. So for a pseudo position of 2, i would have to move the motor at 5?
```
In [13]: ims.forward(2)
Out[13]: FakeSIMOffsetIMSRealPos(motor=5)

In [14]: ims.motor.move(5)
Out[14]: MoveStatus(done=True, pos=fake_ims_motor, elapsed=0.0, success=True, settle_time=0.0)

In [15]: ims
Out[15]: 

fake_ims
--------
position: 0.0

motor
-----
  position: 5

offset
------
  position: 0
  notepad_setpoint: 1
  notepad_readback: 0
```

The offset however does not change, where exactly am I going wrong here?

Reply:
Ohh and i have this guy for testing:

```
class SIMOffsetIMS(OffsetIMS):
    """Simulation of OffsetIMS device for testing."""
    motor = Cpt(FastMotor, limits=(-100, 100))

```

Reply:
I think this works but I don't know how legal this is:

```
    @pseudo_position_argument
    def forward(self, pseudo_pos):
        pseudo_pos = self.PseudoPosition(*pseudo_pos)
        motor_value = pseudo_pos.offset + self.offset.notepad_setpoint.get()
        return self.RealPosition(motor=motor_value)

    @real_position_argument
    def inverse(self, real_pos):
        real_pos = self.RealPosition(*real_pos)
        offset = real_pos.motor - self.offset.notepad_setpoint.get()
        return self.PseudoPosition(offset=offset)
```

Reply:
I think one of us ophyd experts should work on providing a generalized "`PseudoPositioner` with a user offset" - it's something that _should_ be easy but isn't just yet. `DelayBase` is probably the closest thing so far: https://github.com/pcdshub/pcdsdevices/blob/b0a85fb42effa63bf29e3c8f9b668671f6e60df3/pcdsdevices/pseudopos.py#L229

That may mean a further a delay in getting this PR in, though, if that's acceptable...

Reply:
I think it is: we are nearly done with running in 2020 and there is now the target alignment & shooting project to work on as well.

Reply:
I am sorry and thank you. 🙏